backend/src/services/teams/teams.service.js

```diff
@@ -0,0 +1,176 @@
+'use strict';
+
+/**
+ * Teams Service
+ *
+ * Business logic for team operations.
+ */
+
+class TeamsService {
+  constructor(teamsRepo, teamsPermissions) {
+    this.teamsRepo = teamsRepo;
+    this.teamsPermissions = teamsPermissions;
+  }
+
+  async assertCanView(user, course, enrollment) {
+    if (enrollment === null) {
+      const canView = await this.teamsPermissions.canViewTeams(
+        user.id,
+        course.id
+      );
+      if (!canView) {
+        const e = new Error('You are not enrolled in this course');
+        e.code = 'FORBIDDEN';
+        throw e;
+      }
+    }
+  }
+
+  async assertCanModify(user, course, enrollment, action) {
+    if (enrollment === null) {
+      const canModify = await this.teamsPermissions.canModifyTeams(
+        user.id,
+        course.id
+      );
+      if (!canModify) {
+        const e = new Error(
+          action ||
+            'Only professors and TAs can create or update teams / membership'
+        );
+        e.code = 'FORBIDDEN';
+        throw e;
+      }
+    } else if (enrollment.role !== 'professor' && enrollment.role !== 'ta') {
+      const e = new Error(
+        action ||
+          'Only professors and TAs can create or update teams / membership'
+      );
+      e.code = 'FORBIDDEN';
+      throw e;
+    }
+  }
```

Comment:
[Suggestion]
Since we load enrollment information in a preHandler, if enrollment is null, we can say that the user is not enrolled in the course. Requerying the DB would not help here.
Instead, I think here we can simply pass the enrollment object (and other necessary objects) to the permissions repo and the check for role being professor or ta can happen inside there. Right now, that logic is living in permissions and service.

Reply:
[Blocker]
Update the calling code here to reflect the changes in the permissions functions.